pkg/querysharding/analyzer.go

```diff
@@ -5,41 +5,92 @@ package querysharding
 
 import (
 	"fmt"
+	"os"
 
+	"github.com/go-kit/log"
+	lru "github.com/hashicorp/golang-lru"
 	"github.com/prometheus/prometheus/promql/parser"
 )
 
 // QueryAnalyzer is an analyzer which determines
 // whether a PromQL Query is shardable and using which labels.
 type QueryAnalyzer struct{}
 
+var LRU *lru.Cache
+var logger log.Logger
```

Comment:
Global variable is not a good idea here, should be in some structs.

Reply:
What about using a global struct object instead of global variables ?

Reply:
No, we should avoid it if applicable.